src/main/java/services/impl/RestoreServiceImpl.java

```diff
@@ -27,13 +41,74 @@
         RestTemplate restTemplate = new RestTemplate();
         log.info("Restoring at domain: " + domain + " with auth: " + auth);
-        ResponseEntity<String> response =
-                restTemplate.exchange(getRestoreUrl(domain),
-                        HttpMethod.GET,
-                        new HttpEntity<String>(auth.getAuthHeaders()), String.class);
+        HttpHeaders headers = auth.getAuthHeaders();
+        headers.add("x-openrosa-version",  "2.0");
+
+        ResponseEntity<String> response = restTemplate.exchange(
+                getRestoreUrl(domain),
+                HttpMethod.GET,
+                new HttpEntity<String>(headers),
+                String.class
+        );
+        if (response.getStatusCode().value() == 202) {
+            handleAsyncRestoreResponse(response.getBody());
+        }
         return response.getBody();
     }
 
+    /**
+     * Given an async restore xml response, this function throws an AsyncRetryException
+     * with meta data about the async restore.
+     *
+     * @param xml - Async restore response
+     */
+    private void handleAsyncRestoreResponse(String xml) {
+        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+        DocumentBuilder builder;
+        ByteArrayInputStream input;
+        Document doc;
+
+        // Create the XML Document builder
+        try {
+            builder = factory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            throw new RuntimeException("Unable to instantiate document builder");
+        }
+
+        // Parse the xml into a utf-8 byte array
+        try {
+            input = new ByteArrayInputStream(xml.getBytes("utf-8") );
+        } catch (UnsupportedEncodingException e) {
+            throw new RuntimeException("Unable to parse async restore response.");
+        }
+
+        // Build an XML document
+        try {
+            doc = builder.parse(input);
+        } catch (SAXException e) {
+            throw new RuntimeException("Unable to parse into XML Document");
+        } catch (IOException e) {
+            throw new RuntimeException("Unable to parse into XML Document");
+        }
+
+        NodeList messageNodes = doc.getElementsByTagName("message");
+        NodeList progressNodes = doc.getElementsByTagName("progress");
+
+        assert messageNodes.getLength() == 1;
+        assert progressNodes.getLength() == 1;
+
+        String message = messageNodes.item(0).getTextContent();
+        Node progressNode = progressNodes.item(0);
+        NamedNodeMap attributes = progressNode.getAttributes();
+
+        throw new AsyncRetryException(
+                message,
+                Integer.parseInt(attributes.getNamedItem("done").getTextContent()),
+                Integer.parseInt(attributes.getNamedItem("total").getTextContent()),
+                Integer.parseInt(attributes.getNamedItem("retry-after").getTextContent())
```

Comment:
I might have forgotten to mention this, can you use the Rety-After header to get this value instead of reading it from the XML? The mobile does it that way, so would be good keep consistent.